sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -128,8 +128,10 @@ object Cast {
     case _ if from == to => true
     case (from: NumericType, to: DecimalType) if to.isWiderThan(from) => true
     case (from: DecimalType, to: NumericType) if from.isTighterThan(to) => true
-    case (f, t) if legalNumericPrecedence(f, t) => true
+    case (f: NumericType, t: NumericType) if legalNumericPrecedence(f, t) => true
+
     case (DateType, TimestampType) => true
+    case (NullType, _) => false
```

Comment:
Why can't we upcast null to other nullable types? I think it's pretty to write `INSERT INTO tbl VALUES (1, null)`

Reply:
Here we can't know the nullability of the `to` type. We should consider it is not nullable.
For the case you mentioned, it is handled in https://github.com/apache/spark/pull/24806/files#diff-86e655772e8f7cab055d2c2451b52275R134.

Reply:
I agree with @cloud-fan that this should be allowed. Nullability is an additional check, but the types are compatible.